### PR TITLE
R Support [WIP]

### LINKED ISSUE #306: Initial R Support in Grayskull
**Motivation**
Grayskull presently supports Python packages available on PyPI and GitHub. Given the popularity of R in the conda ecosystem, the next obvious step in Grayskull development would be to add support for R packages available on CRAN. This will be a step forward in making Grayskull more versatile and therefore the de facto recipe generator tool for conda.

**Alternatives**
A short-term alternative suggested was to add conda-build as an optional dependency in Grayskull. So that `grayskull cran <package-name>` would run conda-skeleton under the hood and generate the R recipe. This would give a better user experience as the same recipe generator could generate Python and R recipes. But further discussions brought out that adding conda-build as a dependency in Grayskull would defeat the very purpose of creating Grayskull (a stand-alone, conda-build independent application) in the first place. A warning message telling the user that Grayskull doesn't yet support CRAN packages along with a suggestion to use conda-skeleton to generate the R recipe might be the way to go while we work on adding R support to Grayskull. 

**Specifications**

1. For a Python package, the metadata for generating the conda recipe is extracted from setup.py and setup.cfg files. Similarly, an R package is associated with its 'DESCRIPTION file' which contains important metadata about the package. Upon investigation, it is found that several fields in the conda recipe of an R package can be populated from the DESCRIPTION file itself. 

| Section in DESCRIPTION file | Section in conda recipe it maps to |
| ----- | -----|
| imports | requirements: run: |
| Description | about: summary: |
| Url | about: home: |
| License | about: license_family: |

2. A simple R recipe would be one that doesn't include compilers. `NeedsCompilation: no`.
3. How conda-skeleton is doing things is; it simply goes to https://cran.r-project.org/src/contrib/ and in a list saves all the packages found here and then in this list looks for the package the user mentioned.
4. Cannot use https://github.com/ropensci/codemetar because it has a GPL license, also rpy2 doesn't work on windows. 
5. Download the R package tar ball, unzip it, and read the description file, and populate the sections of the recipe (mapped in the above table).

**Observations/Yet to implement**

1. Package name in R package recipes has a "**r-**" prefix. [See example](https://github.com/conda-forge/r-future-feedstock/blob/main/recipe/meta.yaml#L6).
2. The 'orig_lines' key in the Description file has a list of items as value. This list contains some metadata about the package. This is added at the bottom of each conda R recipe in the form of comments. [See example](https://github.com/conda-forge/r-future-feedstock/blob/main/recipe/meta.yaml#L59).
3. The R packages mentioned in the run, build, host sections of the recipe also have the "r-" suffix before their names. [See example](https://github.com/conda-forge/r-future-feedstock/blob/main/recipe/meta.yaml#L27).

**NOTE:** 
This issue will be used as the place for community discussion on how to pull this off. 
As decisions are made and the development work progresses, this tech spec will be updated.

### PR DESCRIPTION
**Pull request by ForgottenProgramme**
Fixes #306 

Add CRAN origin to Grayskull. 
These changes only aim to handle simple R recipes. 
The functionality to extract compiler information from R metadata and therefore handle complex packages will be added in consequent iterations.

ToDo

- [ ] Write unit tests
- [x] Extract and add License
- [x] cran metadata in commented form at the bottom [I am able to add it at the top. not at the bottom]
- [x] r prefix in package name at top
- [x] source: section of the recipe (urls and sha256)
- [x] requirements: host and requirements: build
- [x] Hardcode rpaths:
- [x] set posix and set native at top
- [x] Add r-base to host and run requirements everytime
- [x] reorder the requirements to be sorted. NOTE:
(It is not a requirement to have the packages listed in a sorted order. But it's a nice to have, as it improves readability for users trying to look for a specific package in the list)
- [ ] build:
- [ ] requirements: build:
- [ ] in source: section, the 'url' needs fixing

### CODE REVIEW

**Comment by ForgottenProgramme** [comment]
image

**Comment by ForgottenProgramme** [comment]
@marcelotrevisani

File: grayskull/strategy/cran.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+            else:
+                # Sometimes fields are included but left blank, e.g.:
+                #   - Enhances in data.tree
+                #   - Suggests in corpcor
+                (k, v) = line.split(":", 1)
+        except ValueError:
+            sys.exit("Error: Could not parse metadata (%s)" % line)
+        d[k] = v
+        # if k not in CRAN_KEYS:
+        #     print("Warning: Unknown key %s" % k)
```
Comment (marcelotrevisani) [inline grayskull/strategy/cran.py:64]:
```suggestion
```

File: grayskull/strategy/cran.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                    fp = zf.open(member, "r")
+                    return read_description_contents(fp)
+    else:
+        sys.exit("Cannot extract a DESCRIPTION from file %s" % path)
+    sys.exit("%s does not seem to be a CRAN package (no DESCRIPTION) file" % path)
+
+
+def get_cran_archive_versions(cran_url, session, package, verbose=True):
+    if verbose:
+        print(f"Fetching archived versions for package {package} from {cran_url}")
```
Comment (marcelotrevisani) [inline grayskull/strategy/cran.py:196]: Can you use log, or can you use ``print_msg`` from ``grayskull.cli.stdout`` please?

File: tests/cli/test_cli_cmds.py [hunk 3]
```diff
@@ -103,10 +103,12 @@ def mock_is_pkg_available(pkg):
     assert spy.call_args_list[1].args[0] == {"colorama"}
 
 
-def test_part_reload_recipe(tmpdir):
-    recipe = GrayskullFactory.create_recipe(
-        "pypi", Configuration(name="pytest", version="5.3.2")
-    )
+# The CRAN test is not passing. I don't understand why!
```
Comment (marcelotrevisani) [inline tests/cli/test_cli_cmds.py:106]: you also need to change line ``124`` to 
```
cli.main([index, str(recipe_path), "--sections", "requirements"])
```

**Comment by marcelotrevisani** [comment]
Closing this in favor of https://github.com/conda-incubator/grayskull/pull/423

Thank you very much for your hard work and all effort that you put into this!

**Pull request closed without merging**